File: lib/ts/src/internal/schema/GetTypeUnexpectedParseFailure.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { SchemaParseFailure } from '../../internal/schema/SchemaParseFailure.js';
import { getType } from '../../internal/types/GetType.js';

export function getTypeUnexpectedParseFailure(
    documentName: string,
    path: any[],
    value: any,
    expectedType: string,
): SchemaParseFailure[] {
    const actualType = getType(value);
    const data: { actual: { [key: string]: any }; expected: { [key: string]: any } } = {
        actual: { [actualType]: {} },
        expected: { [expectedType]: {} },
    };
    return [new SchemaParseFailure(documentName, path, 'TypeUnexpected', data)];
}

File: lib/go/internal/schema/GetTypeUnexpectedParseFailure.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package schema

import "github.com/telepact/telepact/lib/go/internal/types"

// GetTypeUnexpectedParseFailure constructs a parse failure describing an unexpected type.
func GetTypeUnexpectedParseFailure(documentName string, path []any, value any, expectedType string) []*SchemaParseFailure {
	actualType := types.GetType(value)
	data := map[string]any{
		"actual":   map[string]any{actualType: map[string]any{}},
		"expected": map[string]any{expectedType: map[string]any{}},
	}

	return []*SchemaParseFailure{
		NewSchemaParseFailure(documentName, path, "TypeUnexpected", data),
	}
}
